components/builder-graph/src/package_graph_trait.rs

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) 2017 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::protocol::originsrv;
+
+#[derive(Debug)]
+pub struct Stats {
+    pub node_count:     usize,
+    pub edge_count:     usize,
+    pub connected_comp: usize,
+    pub is_cyclic:      bool,
+}
+
+pub trait PackageGraphTrait: Send + Sync {
+    fn build(&mut self,
+             packages: &[originsrv::OriginPackage],
+             use_build_deps: bool)
+             -> (usize, usize);
+    fn extend(&mut self,
+              package: &originsrv::OriginPackage,
+              use_build_deps: bool)
+              -> (usize, usize);
+    fn check_extend(&mut self, package: &originsrv::OriginPackage, use_build_deps: bool) -> bool;
```

Comment:
This is a pretty high level operation. I wonder if it could be expressed genericly as a set of other graph operations (eg `contains`)? This would allow `check_extend` to have a default implementation. This could potentially reduce effort and code duplication as more things implement this trait.

Reply:
We made an implementation decision to try and slide the trait in first, then address later some of the type and shape issues with the trait. 
In this case there will only be the old and new graph code here, and hopefully only the new code long term. The trait becomes a little bit redundant at that point, like scaffolding left up after a refurbishing.